ott/tools/k_means.py

```diff
@@ -0,0 +1,366 @@
+          f"Expected initial centroids to have shape "
+          f"`{k, geom.cost_rank}`, found `{centroids.shape}`."
+      )
+    n = geom.shape[0]
+    prev_assignment = jnp.full((n,), -2)
+    assignment = jnp.full((n,), -1)
+    errors = jnp.full((max_iterations,), -1.)
+
+    return KMeansState(
+        centroids=centroids,
+        prev_assignment=prev_assignment,
+        assignment=assignment,
+        center_shift=jnp.inf,
+        errors=errors,
+    )
+
+  def cond_fn(iteration: int, const: Any, state: KMeansState) -> bool:
+    del const
+    assignment_not_same = jnp.any(state.prev_assignment != state.assignment)
+    tol_not_satisfied = state.center_shift > tol
+    return jnp.logical_and(assignment_not_same, tol_not_satisfied)
+
+  def body_fn(
+      iteration: int, const: Any, state: KMeansState, compute_error: bool
+  ) -> KMeansState:
+    del compute_error, const
+
+    assignment, dist_to_centers = update_assignment(state.centroids)
+    centroids = update_centroids(assignment, dist_to_centers)
+    err = jnp.sum(weights * dist_to_centers)
+    errors = state.errors.at[iteration].set(err)
+
+    return KMeansState(
+        centroids=centroids,
+        prev_assignment=state.assignment,
+        assignment=assignment,
+        center_shift=center_shift(state.centroids, centroids),
+        errors=errors,
+    )
+
+  def finalize_assignment(state: KMeansState) -> KMeansState:
+    last_iter = jnp.sum(state.errors != -1) - 1
+    assignment, dist_to_centers = update_assignment(state.centroids)
+    err = jnp.sum(weights * dist_to_centers)
+    return state._replace(
+        assignment=assignment, errors=state.errors.at[last_iter].set(err)
+    )
+
+  weighted_x = jnp.hstack([weights[:, None] * geom.x, weights[:, None]])
+  state = fixed_point_loop.fixpoint_iter(
+      cond_fn,
+      body_fn,
+      min_iterations=min_iterations,
+      max_iterations=max_iterations,
+      inner_iterations=1,
+      constants=None,
+      state=init_fn(init)
+  )
+  state = jax.lax.cond(
+      jnp.all(state.prev_assignment == state.assignment), lambda _: _,
+      finalize_assignment, state
+  )
+
+  return KMeansOutput._from_state(
+      state, tol=tol, store_inner_errors=store_inner_errors
+  )
+
+
+def k_means(
+    geom: Union[jnp.ndarray, pointcloud.PointCloud],
+    k: int,
+    weights: Optional[jnp.ndarray] = None,
+    init: Init_t = "k-means++",
+    n_init: int = 10,
+    n_local_trials: Optional[int] = None,
+    tol: float = 1e-4,
+    min_iterations: int = 0,
+    max_iterations: int = 300,
+    store_inner_errors: bool = False,
+    key: Optional[jnp.ndarray] = None,
+) -> KMeansOutput:
+  r"""K-means clustering using Lloyd's algorithm :cite:`lloyd:82`.
+
+  Args:
+    geom: Point cloud of shape ``[n, ndim]`` to cluster. If passed as an array,
+      :class:`~ott.geometry.costs.Euclidean` cost is assumed.
+    k: The number of clusters.
+    weights: The weights of each point. If ``None``, uniform weights are used.
```

Comment:
maybe this is really just a detail, but maybe better to specify, since weights are also sometimes considered for centroids, that those are weights for _input_ points.

Reply:
Done, thanks!